File: app/src/main/java/org/stepic/droid/store/dao/IDao.java
package org.stepic.droid.store.dao;

import android.content.ContentValues;
import android.database.Cursor;

import org.jetbrains.annotations.NotNull;

import java.util.List;

public interface IDao<T> {
    void insertOrUpdate(T persistentObject);

    boolean isInDb(T persistentObject);

    boolean isInDb(String whereColumn, String value);

    @NotNull
    List<T> getAll();

    @NotNull
    List<T> getAll(String whereColumnName, String whereValue);

    T get(String whereColumnName, String whereValue);

    T parsePersistentObject(Cursor cursor);

    String getDbName();

    ContentValues getContentValues(T persistentObject);

    String getDefaultPrimaryColumn();

    String getDefaultPrimaryValue(T persistentObject);

    void update(String whereColumn, String whereValue, ContentValues contentValues);

    void delete(String whereColumn, String whereValue);

    List<T>  getAllInRange(String  whereColumn, String commaSeparatedIds);
}

File: app/src/main/java/org/stepic/droid/store/dao/IDao.kt
package org.stepic.droid.store.dao

import android.content.ContentValues

interface IDao<T> {
    fun insertOrUpdate(persistentObject: T)

    fun isInDb(persistentObject: T): Boolean

    fun isInDb(whereColumn: String, value: String): Boolean

    fun getAll():List<T?>

    fun getAll(whereColumnName: String, whereValue: String): List<T?>

    fun get(whereColumnName: String, whereValue: String): T?

    fun update(whereColumn: String, whereValue: String, contentValues: ContentValues)

    fun delete(whereColumn: String, whereValue: String)

    fun getAllInRange(whereColumn: String, commaSeparatedIds: String): List<T>

    @Deprecated("it is hack for two course tables, it will be removed")
    fun setTableName(name: String)
}
